crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -0,0 +1,297 @@
+    /// which depends on transaction [gas_limit].
+    /// Maximum allocation size is 2^32 - 1 bytes;
+    pub fn new(gas_limit: u64) -> Self {
+        let size = min(
+            SharedMemory::calculate_upper_bound(gas_limit) as usize,
+            2usize.pow(32) - 1,
+        );
+
+        let mut data = vec![0; size];
+        let current_slice: *mut [u8] = &mut data[..];
+        SharedMemory {
+            data,
+            checkpoints: Vec::with_capacity(1024),
+            current_slice,
+            current_len: 0,
+            limit: size as u64,
+        }
+    }
+
+    /// Allocate memory to be shared between calls.
+    /// Memory size is estimated using https://2π.com/22/eth-max-mem
+    /// which depends on transaction [gas_limit].
+    /// Uses [memory_limit] as maximum allocation size
+    pub fn new_with_memory_limit(gas_limit: u64, memory_limit: u64) -> Self {
+        let size = min(
+            SharedMemory::calculate_upper_bound(gas_limit) as usize,
+            memory_limit as usize,
+        );
+
+        let mut data = vec![0; size];
+        let current_slice: *mut [u8] = &mut data[..];
+        SharedMemory {
+            data,
+            checkpoints: Vec::with_capacity(1024),
+            current_slice,
+            current_len: 0,
+            limit: size as u64,
+        }
+    }
+
+    /// Prepares the shared memory for a new context
+    pub fn new_context_memory(&mut self) {
+        let base_offset = self.checkpoints.last().unwrap_or(&0);
+        let new_checkpoint = base_offset + self.current_len;
+
+        self.checkpoints.push(new_checkpoint);
+
+        self.current_slice = &mut self.data[new_checkpoint..];
+        self.current_len = 0;
+    }
+
+    /// Prepares the shared memory for returning to the previous context
+    pub fn free_context_memory(&mut self) {
+        if let Some(old_checkpoint) = self.checkpoints.pop() {
+            let last = *self.checkpoints.last().unwrap_or(&0);
+            self.current_slice = &mut self.data[last..];
+            self.current_len = old_checkpoint - last;
+            self.update_limit();
+        }
+    }
+
+    /// Get the length of the current memory range.
+    pub fn len(&self) -> usize {
+        self.current_len
+    }
+
+    /// Return true if current effective memory range is zero.
+    pub fn is_empty(&self) -> bool {
+        self.current_len == 0
+    }
+
+    /// Return the memory of the current context
+    pub fn data(&self) -> &[u8] {
+        self.get_current_slice()
+    }
+
+    /// Resize the memory. assume that we already checked if:
+    /// - we have enought gas to resize this vector
+    /// - we made new_size as multiply of 32
+    /// - [new_size] is greater than `self.len()`
+    pub fn resize(&mut self, new_size: usize) {
+        // extend with zeros
+        let range = self.current_len..new_size;
+
+        self.get_current_slice_mut()[range]
+            .iter_mut()
+            .for_each(|byte| *byte = 0);
```

Comment:
This can be `ptr::write_bytes(self.ptr.add(self.len), 0, new_size)`

Reply:
Ok I get your point regarding `new_size` and `len`. However, this function is always used in the `shared_memory_resize` macro where this check is already done. If you want, I can move that check inside here and remove it from the macro

Regarding the code snippet you wrote, I'll try to implement it! thanks

Reply:
Ah I see that, we should move the check to this function since it's public (make it return `Result<(), _>` like the stack functions)

Reply:
Sorry for all the questions but
- I tried to move the check inside the resize function, however the check still needs to live inside the macro because of the gas calculation that needs to be skipped, otherwise I get a root mismatch. Is this still convenient to do?
- I'm trying to understand the `data_mut` as suggested from `slice::from_raw_parts_mut(self.ptr, self.data.as_ptr().add(self.data.len()).sub(self.ptr))`. What is `self.ptr` here?

Reply:
- that's fine, you can keep it in the macro then
- this https://github.com/bluealloy/revm/pull/673#discussion_r1337555841

Reply:
Take a look at the `resize` function now, I should have matched what you wanted. The for loop should be faster than calling the closure with its overhead

Reply:
Changed the `for` loop with the `fill` method for consistency